File: csharp/tests/Vanni.Idmix.Tests/CrossLanguageFixtures.cs
using Vanni.Idmix;
using Xunit;

namespace Vanni.Idmix.Tests;

/// <summary>与 testdata/cross_language_vectors.json 同步的跨语言测试向量。</summary>
internal static class CrossLanguageFixtures
{
    internal const string Alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    internal record CrossLangValue(int OType, long Val, string? StrValue)
    {
        internal static CrossLangValue Int(int otype, long val) => new(otype, val, null);
        internal static CrossLangValue Str(string str) => new(0, 0, str);
    }

    internal record Case(string Name, int Variant, string Encoded, CrossLangValue[] Values);

    internal static Case[] Cases { get; } =
    [
        new("spec_example", 0, "ixHjl0FK7",
            [CrossLangValue.Int(1, 5), CrossLangValue.Int(7, -1), CrossLangValue.Int(2, 40)]),
        new("uint32_max", 0, "hUdZLNKGa", [CrossLangValue.Int(2, 4294967295L)]),
        new("int32_min", 0, "hUdElRoHP", [CrossLangValue.Int(6, -2147483648)]),
        new("int64_min", 0, "8B10qg6x0EAf3b", [CrossLangValue.Int(7, long.MinValue)]),
        new("int64_max", 0, "8B2cU8kbWpQ2RM", [CrossLangValue.Int(7, long.MaxValue)]),
        new("uint64_max", 0, "8B3CPRsv0Owa6S", [CrossLangValue.Int(3, unchecked((long)ulong.MaxValue))]),
        new("mixed_extremes", 0, "bULoRnNZJinEZGKD78wIigIaw6QplS8B0HGNCKO2L6",
            [
                CrossLangValue.Int(2, 4294967295L),
                CrossLangValue.Int(6, -2147483648),
                CrossLangValue.Int(7, long.MinValue),
                CrossLangValue.Int(7, long.MaxValue),
            ]),
        new("embedded_small", 0, "ixHorRWmh",
            [
                CrossLangValue.Int(0, 15),
                CrossLangValue.Int(4, -16),
                CrossLangValue.Int(1, 0),
                CrossLangValue.Int(5, -1),
            ]),
        new("access_key", 0, "eNe8RmcNtYw60Xjc",
            [CrossLangValue.Int(2, 1001), CrossLangValue.Int(3, 1690000000L), CrossLangValue.Int(0, 3)]),
        new("string_example", 0, "ceOqw5RPaTfgnfXyp7Sdepb",
            [CrossLangValue.Str("hello"), CrossLangValue.Int(1, 5), CrossLangValue.Str("世界")]),
    ];

    internal static object Materialize(CrossLangValue v)
    {
        if (v.StrValue != null) return v.StrValue;
        return v.OType switch
        {
            TypedValue.OTypeUint8 => (byte)v.Val,
            TypedValue.OTypeUint16 => (ushort)v.Val,
            TypedValue.OTypeUint32 => (uint)v.Val,
            TypedValue.OTypeUint64 => TypedValue.U64(unchecked((ulong)v.Val)),
            TypedValue.OTypeInt8 => (sbyte)v.Val,
            TypedValue.OTypeInt16 => (short)v.Val,
            TypedValue.OTypeInt32 => (int)v.Val,
            TypedValue.OTypeInt64 => v.Val,
            _ => throw new ArgumentException($"invalid otype {v.OType}"),
        };
    }

    internal static void AssertValueEquals(CrossLangValue want, object got, int index)
    {
        if (want.StrValue != null)
        {
            Assert.Equal(want.StrValue, Assert.IsType<string>(got));
            return;
        }
        var obj = Number.ObjectFromAny(got);
        Assert.False(obj.IsString);
        Assert.Equal(want.OType, obj.OType);
        Assert.Equal(want.Val, obj.Val);
    }
}

File: java/src/test/java/io/github/vannifan/idmix/CrossLanguageFixtures.java
package io.github.vannifan.idmix;

import java.util.List;

/** Cross-language test vectors synced with testdata/cross_language_vectors.json. */
final class CrossLanguageFixtures {
    private CrossLanguageFixtures() {}

    static final String ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    record CrossLangValue(int otype, long val, String str) {
        static CrossLangValue integer(int otype, long val) { return new CrossLangValue(otype, val, null); }
        static CrossLangValue str(String str) { return new CrossLangValue(0, 0, str); }
    }

    record Case(String name, int variant, String encoded, List<CrossLangValue> values) {}

    static List<Case> cases() {
        return List.of(
                new Case("spec_example", 0, "ixHjl0FK7",
                        List.of(
                                CrossLangValue.integer(1, 5),
                                CrossLangValue.integer(7, -1),
                                CrossLangValue.integer(2, 40))),
                new Case("uint32_max", 0, "hUdZLNKGa",
                        List.of(CrossLangValue.integer(2, 4294967295L))),
                new Case("int32_min", 0, "hUdElRoHP",
                        List.of(CrossLangValue.integer(6, -2147483648))),
                new Case("int64_min", 0, "8B10qg6x0EAf3b",
                        List.of(CrossLangValue.integer(7, Long.MIN_VALUE))),
                new Case("int64_max", 0, "8B2cU8kbWpQ2RM",
                        List.of(CrossLangValue.integer(7, Long.MAX_VALUE))),
                new Case("uint64_max", 0, "8B3CPRsv0Owa6S",
                        List.of(CrossLangValue.integer(3, Long.parseUnsignedLong("18446744073709551615")))),
                new Case("mixed_extremes", 0, "bULoRnNZJinEZGKD78wIigIaw6QplS8B0HGNCKO2L6",
                        List.of(
                                CrossLangValue.integer(2, 4294967295L),
                                CrossLangValue.integer(6, -2147483648),
                                CrossLangValue.integer(7, Long.MIN_VALUE),
                                CrossLangValue.integer(7, Long.MAX_VALUE))),
                new Case("embedded_small", 0, "ixHorRWmh",
                        List.of(
                                CrossLangValue.integer(0, 15),
                                CrossLangValue.integer(4, -16),
                                CrossLangValue.integer(1, 0),
                                CrossLangValue.integer(5, -1))),
                new Case("access_key", 0, "eNe8RmcNtYw60Xjc",
                        List.of(
                                CrossLangValue.integer(2, 1001),
                                CrossLangValue.integer(3, 1690000000L),
                                CrossLangValue.integer(0, 3))),
                new Case("string_example", 0, "ceOqw5RPaTfgnfXyp7Sdepb",
                        List.of(
                                CrossLangValue.str("hello"),
                                CrossLangValue.integer(1, 5),
                                CrossLangValue.str("世界")))
        );
    }

    static Object materialize(CrossLangValue v) {
        if (v.str() != null) return v.str();
        return switch (v.otype()) {
            case TypedValue.OTYPE_UINT8 -> TypedValue.u8((int) v.val());
            case TypedValue.OTYPE_UINT16 -> TypedValue.u16((int) v.val());
            case TypedValue.OTYPE_UINT32 -> TypedValue.u32(v.val());
            case TypedValue.OTYPE_UINT64 -> TypedValue.u64(Long.toUnsignedString(v.val()));
            case TypedValue.OTYPE_INT8 -> TypedValue.i8((int) v.val());
            case TypedValue.OTYPE_INT16 -> TypedValue.i16((int) v.val());
            case TypedValue.OTYPE_INT32 -> TypedValue.i32((int) v.val());
            case TypedValue.OTYPE_INT64 -> TypedValue.i64(v.val());
            default -> throw new IllegalArgumentException("invalid otype " + v.otype());
        };
    }

    static void assertValueEquals(CrossLangValue want, Object got, String label) {
        if (want.str() != null) {
            if (!(got instanceof String s) || !want.str().equals(s)) {
                throw new AssertionError(label + ": got str=" + got + ", want " + want.str());
            }
            return;
        }
        if (got instanceof TypedValue tv) {
            if (tv.otype != want.otype() || tv.val != want.val()) {
                throw new AssertionError(label + ": got otype=" + tv.otype + " val=" + tv.val
                        + ", want otype=" + want.otype() + " val=" + want.val());
            }
            return;
        }
        Number.DataObject gotObj = Number.objectFromAny(got);
        if (gotObj.isString || gotObj.otype != want.otype() || gotObj.val != want.val()) {
            throw new AssertionError(label + ": got otype=" + gotObj.otype + " val=" + gotObj.val
                    + ", want otype=" + want.otype() + " val=" + want.val());
        }
    }
}
